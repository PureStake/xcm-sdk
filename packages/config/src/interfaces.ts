// eslint-disable-next-line import/no-cycle
import { Asset, Chain, MoonChain } from './constants';

export interface AssetConfig<Assets extends Asset = Asset> {
  id: string;
  /**
   * id -> erc20Id
   * `0xffffffff${BigInt(id).toString(16).padStart(32, '0')}`
   * exceptions are native tokens, for them we use
   * Erc20BalancesPrecompile 0x0000000000000000000000000000000000000802
   */
  erc20Id: string;
  originSymbol: Assets;
  originAssetId?: number;
  isNative?: boolean;
}

export interface ChainBaseConfig<Chains = Chain> {
  chain: Chains;
  name: string;
  ws: string;
  parachainId: number;
}

export interface MoonChainConfig extends ChainBaseConfig<MoonChain> {
  decimals: number;
}

export interface ChainConfig extends ChainBaseConfig {
  weight: number;
  moonAssetId?: number | bigint;
  palletInstance?: number;
}
