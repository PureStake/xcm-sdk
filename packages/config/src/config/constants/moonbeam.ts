import { createBalanceBuilder } from '../../balance';
import {
  Assets,
  Chain,
  MOONBEAM_ASSETS_CONFIGS as assets,
  MOONBEAM_CHINS_CONFIGS as chains,
  MoonChain,
  MOON_CHINS_CONFIGS,
} from '../../constants';
import {
  createExtrinsicBuilder,
  XTokensExtrinsicSuccessEvent,
} from '../../extrinsic';
import { MoonbeamAssets } from '../../interfaces';
import { createWithdrawBuilder } from '../../withdraw';
import { ChainXcmConfigs } from '../config.interfaces';

const config = MOON_CHINS_CONFIGS[MoonChain.Moonbeam];
const balance = createBalanceBuilder<MoonbeamAssets>();
const extrinsic = createExtrinsicBuilder<MoonbeamAssets>(config);
const withdraw = createWithdrawBuilder<MoonbeamAssets>();

export const MOONBEAM_CONFIGS: ChainXcmConfigs<MoonbeamAssets> = {
  [Assets.ACA]: {
    asset: assets[Assets.ACA],
    origin: chains[Chain.Acala],
    deposit: [
      {
        balance: balance.system(),
        extrinsic: extrinsic
          .xTokens()
          .transfer()
          .successEvent(XTokensExtrinsicSuccessEvent.TransferredMultiAssets)
          .params(chains[Chain.Acala], {
            Token: assets[Assets.ACA].originSymbol,
          }),
        origin: chains[Chain.Acala],
      },
    ],
    withdraw: [
      withdraw.xTokens({
        balance: balance.system(),
        destination: chains[Chain.Acala],
        existentialDeposit: 100_000_000_000,
        feePerWeight: 8,
      }),
    ],
  },
  [Assets.AUSD]: {
    asset: assets[Assets.AUSD],
    origin: chains[Chain.Acala],
    deposit: [],
    withdraw: [],
  },
  [Assets.DOT]: {
    asset: assets[Assets.DOT],
    origin: chains[Chain.Polkadot],
    deposit: [],
    withdraw: [],
  },
  [Assets.GLMR]: {
    asset: assets[Assets.GLMR],
    origin: config,
    deposit: [],
    withdraw: [],
  },
  [Assets.PARA]: {
    asset: assets[Assets.PARA],
    origin: chains[Chain.Parallel],
    deposit: [],
    withdraw: [],
  },
};
