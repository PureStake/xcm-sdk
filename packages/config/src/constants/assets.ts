export enum Asset {
  ACA = 'ACA',
  ASTR = 'ASTR',
  AUSD = 'AUSD',
  BNC = 'BNC',
  BSX = 'BSX',
  CRAB = 'CRAB',
  CRU = 'CRU',
  CSM = 'CSM',
  DEV = 'DEV',
  DOT = 'DOT',
  GLMR = 'GLMR',
  HKO = 'HKO',
  KAR = 'KAR',
  KBTC = 'KBTC',
  KINT = 'KINT',
  KMA = 'KMA',
  KSM = 'KSM',
  KUSD = 'KUSD',
  LIT = 'LIT',
  MOVR = 'MOVR',
  PARA = 'PARA',
  PARING = 'PARING',
  PHA = 'PHA',
  RMRK = 'RMRK',
  SDN = 'SDN',
  TEER = 'TEER',
  UNIT = 'UNIT',
  USDT = 'USDT',
}
