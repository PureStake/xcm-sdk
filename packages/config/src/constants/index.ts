/* eslint-disable import/no-cycle */
export * from './assets';
export * from './chains';
