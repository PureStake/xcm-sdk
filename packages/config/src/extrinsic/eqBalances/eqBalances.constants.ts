export enum EqBalancesExtrinsic {
  XcmTransfer = 'xcmTransfer',
  TransferXcm = 'transferXcm',
}

export enum EqBalancesSuccessEvent {
  ExtrinsicSuccess = 'ExtrinsicSuccess',
}

export enum EqBalancesFee {
  SovereignAccWillPay = 'SovereignAccWillPay',
  TargetChainAccWillPay = 'TargetChainAccWillPay',
  ThisAccWillPay = 'ThisAccWillPay',
}
