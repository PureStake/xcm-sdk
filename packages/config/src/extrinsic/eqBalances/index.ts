export * from './eqBalances';
export * from './eqBalances.constants';
export * from './eqBalances.interfaces';
