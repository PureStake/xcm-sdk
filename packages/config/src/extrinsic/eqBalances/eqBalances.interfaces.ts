import { XcmExtrinsicGetParams } from '../common.interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  EqBalancesExtrinsic,
  EqBalancesFee,
  EqBalancesSuccessEvent,
} from './eqBalances.constants';

export interface EqBalancesPallet {
  pallet: ExtrinsicPallet.EqBalances;
  extrinsic: EqBalancesExtrinsic;
  successEvent: EqBalancesSuccessEvent;
  getParams: (params: XcmExtrinsicGetParams) => EqBalancesPalletParams;
}

export type EqBalancesPalletParams = [
  /**
   * asset
   */
  number,
  /**
   * amount
   */
  bigint,
  /**
   * destination
   */
  {
    parents: 1;
    interior: {
      X2: [
        {
          Parachain: number;
        },
        {
          AccountKey20: {
            network: 'Any';
            /**
             * account
             */
            key: string;
          };
        },
      ];
    };
  },
  /**
   * Fee type
   */
  EqBalancesFee,
];
