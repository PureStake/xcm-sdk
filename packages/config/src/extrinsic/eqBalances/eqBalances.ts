import { MoonChain } from '../../interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  EqBalancesExtrinsic,
  EqBalancesSuccessEvent,
} from './eqBalances.constants';
import { EqBalancesPallet } from './eqBalances.interfaces';

/* eslint-disable @typescript-eslint/no-use-before-define */
export function eqBalances(chain: MoonChain) {
  return {
    xcmTransfer: () => xcmTransfer(chain),
  };
}

function xcmTransfer(chain: MoonChain) {
  return {
    successEvent: (event: EqBalancesSuccessEvent) => ({
      asset: (id: number): EqBalancesPallet => ({
        pallet: ExtrinsicPallet.EqBalances,
        extrinsic: EqBalancesExtrinsic.XcmTransfer,
        successEvent: event,
        getParams: ({ account, amount }) => [
          id,
          amount,
          {
            parents: 1,
            interior: {
              X2: [
                {
                  Parachain: chain.parachainId,
                },
                {
                  AccountKey20: {
                    network: 'Any',
                    key: account,
                  },
                },
              ],
            },
          },
          'SovereignAccWillPay',
        ],
      }),
    }),
  };
}
