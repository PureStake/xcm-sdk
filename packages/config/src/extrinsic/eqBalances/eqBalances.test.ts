import { MoonChainKey } from '../../constants';
import { MoonChain } from '../../interfaces';
import { eqBalances } from './eqBalances';
import { EqBalancesFee, EqBalancesSuccessEvent } from './eqBalances.constants';

describe('eqBalances', () => {
  const amount = 1000n;
  const account = '0xeF46c7649270C912704fB09B75097f6E32208b85';
  const eqId = 25_969;
  const chain: MoonChain = {
    key: MoonChainKey.MoonbaseAlpha,
    name: 'Moonbase Alpha',
    ws: 'wss://wss.api.moonbase.moonbeam.network',
    parachainId: 1000,
    decimals: 10,
    chainId: 1287,
    unitsPerSecond: 100n,
  };

  const extrinsic = eqBalances(chain);

  describe('xcmTransfer', () => {
    const cfg = extrinsic
      .xcmTransfer()
      .successEvent(EqBalancesSuccessEvent.ExtrinsicSuccess)
      .asset(eqId)
      .fee(EqBalancesFee.SovereignAccWillPay);

    it('should be correct config', () => {
      expect(cfg).toMatchSnapshot();
    });

    it('should get correct params', () => {
      expect(cfg.getParams({ account, amount })).toMatchSnapshot();
    });
  });
});
