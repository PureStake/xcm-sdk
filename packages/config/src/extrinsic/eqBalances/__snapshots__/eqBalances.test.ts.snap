// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eqBalances xcmTransfer should be correct config 1`] = `
{
  "extrinsic": "xcmTransfer",
  "getParams": [Function],
  "pallet": "eqBalances",
  "successEvent": "ExtrinsicSuccess",
}
`;

exports[`eqBalances xcmTransfer should get correct params 1`] = `
[
  25969,
  1000n,
  {
    "interior": {
      "X2": [
        {
          "Parachain": 1000,
        },
        {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": "Any",
          },
        },
      ],
    },
    "parents": 1,
  },
  "SovereignAccWillPay",
]
`;
